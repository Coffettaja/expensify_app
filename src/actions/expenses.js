import uuid from 'uuid'
import database from '../firebase/firebase'

//  The action generators for expenses

// Add expense
// Takes an object as an argument. The object will be used to set the expense properties. Default values for the argument object and default properties also set. An expense gets an ID value from uuid library function. 
// export const addExpense = (
//   {
//     description = '',
//     note = '',
//     amount = 0,
//     createdAt = 0
//   } = {}
// ) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt,
//   }
// })

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

// starts the process of adding an expense to the store
export const startAddExpense = (expenseData = {}) => {
  // returning functions is OK because of redux-thunk middleware
  // it also gives them the dispatch function as a parameter
  return (dispatch) => {
    // this has basically the same result as default values in addExpense but maybe easier to read
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData

    // First push the data to the database, and if that works, dispatch the action to the redux store
    const expense = { description, note, amount, createdAt }
    // returning a promise to make promise chaining possible (mainly for testing purposes)
    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })
  }
}

/**
 * Generates an action object that states what expense should be removed by ID
 * @return {Object} an action object for removing expenses from the state
 * @param {string} id
 */
export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
})

/**
 * Generates an action object with id of expense to be edited and updates.
 *
 * @param {string} id of the expense to be edited
 * @param {Object} updates - states the new property values
 * @param {string} [updates.description] - new description
 * @returns {Object} with the type: 'EDIT_EXPENSE'
 */
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})