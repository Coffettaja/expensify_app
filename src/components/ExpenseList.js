import React from 'react'
import { connect } from 'react-redux' // connects the component to the redux store
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

// exported for testing purposes
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    {
      props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <p>No expenses</p>
        </div>
      ) : (
        <div>
          {props.expenses.map((expense) => (
            <ExpenseListItem
              key={expense.id}
              {...expense}
            ></ExpenseListItem>
          ))}
        </div>
      )
    }
    
  </div>
)

// Maps the component state to props.
// This automatically reruns when the redux state changes.
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

// The argument is a function that is used to provide what info we want the component to be able to access. The function takes the state of the store as an argument, and returns an object with key - value pairs. The object then gets passed to the connected component as props.
export default connect(mapStateToProps)(ExpenseList)
