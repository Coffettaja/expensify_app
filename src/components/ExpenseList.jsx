import React from 'react'
import { connect } from 'react-redux' // connects the component to the redux store

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.length}
    {props.filters.text}
  </div>
)

// Maps the component state to props.
// This automatically reruns when the redux state changes.
const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    filters: state.filters,
  }
}

// The argument is a function that is used to provide what info we want the component to be able to access. The function takes the state of the store as an argument, and returns an object with key - value pairs. The object then gets passed to the connected component as props.
export default connect(mapStateToProps)(ExpenseList)
