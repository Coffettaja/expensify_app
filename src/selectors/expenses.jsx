import moment from 'moment'
import { create } from 'domain';

// Gets the visible expenses in sorted order. Takes two arguments: expenses, which is the complete array of expenses, and filters that are used to filter the expenses shown.
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt)
      // If startDate is set on the calendar, check if the startDate is earlier than the createdAt of the expense, by day.
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
      const textMatch = expense.description.toLowerCase().includes  (text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch
    })
    // Sorts the expenses by either date or amount
    .sort((a, b) => {
      if (sortBy === 'date') {
        return b.createdAt - a.createdAt
      }
      if (sortBy === 'amount') {
        // When sorting by amount, the highest amount obviously comes first.
        // However, if the amounts are equal, the newer expense comes first.
        if (b.amount - a.amount !== 0) {
          return b.amount - a.amount
        }
        return b.createdAt - a.createdAt
      }
    })
}

export default getVisibleExpenses