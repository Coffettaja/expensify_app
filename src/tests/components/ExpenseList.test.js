import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList.jsx'
import expenses from '../fixtures/expenses'

test('should render ExpenseList with empty message when no expenses', () => {  
  const wrapper = shallow(<ExpenseList expenses={[]}></ExpenseList>)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses}></ExpenseList>)
  expect(wrapper).toMatchSnapshot()
})