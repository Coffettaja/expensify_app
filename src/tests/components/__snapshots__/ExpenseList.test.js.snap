// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message when no expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ul>
    <ExpenseListItem
      amount={6000}
      createdAt={0}
      description="car"
      id={0}
      key="0"
      note=""
    />
    <ExpenseListItem
      amount={16000}
      createdAt={-345600000}
      description="Rent"
      id={1}
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={16000}
      createdAt={345600000}
      description="rent"
      id={2}
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={1000}
      createdAt={1036800000}
      description="water"
      id={3}
      key="3"
      note=""
    />
  </ul>
</div>
`;
